service/src/main/java/bio/terra/workspace/common/utils/GcpUtils.java

```diff
@@ -77,11 +80,26 @@ public static void pollAndRetry(
       throws RetryException, IOException, InterruptedException {
     operation = OperationUtils.pollUntilComplete(operation, pollingInterval, timeout);
     if (operation.getOperationAdapter().getError() != null) {
-      int code = operation.getOperationAdapter().getError().getCode();
-      if (is4xxClientError(code)) {
-        // do not waste time retrying on client error.
-        throw new BadRequestException(
-            String.format("Gcp calls failed with client error code %s. Do not retry", code));
+      int intCode = operation.getOperationAdapter().getError().getCode();
+      // Do not waste time retrying on client error.
+      if (is4xxClientError(intCode)) {
+        Code code = Code.values()[intCode];
+        String errorMessage = operation.getOperationAdapter().getError().getMessage();
+        switch (code) {
+            // 400
+          case INVALID_ARGUMENT, OUT_OF_RANGE, FAILED_PRECONDITION -> throw new BadRequestException(
+              errorMessage);
+            // 401
+          case UNAUTHENTICATED -> throw new UnauthorizedException(errorMessage);
+            // 403
+          case PERMISSION_DENIED -> throw new ForbiddenException(errorMessage);
+            // 409
+          case ALREADY_EXISTS, ABORTED -> throw new ConflictException(errorMessage);
+            // 429
+          case RESOURCE_EXHAUSTED -> throw new BadRequestException(errorMessage);
+          default -> throw new BadRequestException(
```

Comment:
You've already checked for every value of `code` that satisfies `is4xxClientError`, so I don't think you'll ever hit this default case. I think you can just remove `is4xxClientError` and throw the RetryException as the default case